src/renderer/src/components/providers/pinned-tabs-provider.tsx

```
@@ -0,0 +1,124 @@
+    // Optimistically update local state to mask IPC latency.
+    // Mirror the backend logic: set the fractional position, sort, normalize.
+    setPinnedTabsByProfile((prev) => {
+      const next = { ...prev };
+      for (const profileId of Object.keys(next)) {
+        const tabs = next[profileId];
+        const tabIndex = tabs.findIndex((t) => t.uniqueId === pinnedTabId);
+        if (tabIndex === -1) continue;
+
+        const updated = tabs.map((t) => ({ ...t, position: t.uniqueId === pinnedTabId ? newPosition : t.position }));
+        updated.sort((a, b) => a.position - b.position);
+        updated.forEach((t, i) => (t.position = i));
```

Biome (2.4.6)

[error] This callback passed to forEach() iterable method should not return a value.

(lint/suspicious/useIterableCallbackReturn)